windows/main.c

```diff
@@ -0,0 +1,89 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License 2.0;
+// you may not use this file except in compliance with the Elastic License 2.0.
+
+// +build windows
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <windows.h>
+#include "elastic-agent-windows-amd64.h"
+
+SERVICE_STATUS serviceStatus;
+SERVICE_STATUS_HANDLE serviceStatusHandle;
+
+// ServiceCleanup is an atexit handler that sets the service to stopped.
+void ServiceCleanup(void) {
+    if (serviceStatusHandle) {
+        serviceStatus.dwCurrentState = SERVICE_STOPPED;
+        SetServiceStatus(serviceStatusHandle, &serviceStatus);
+    }
+}
+
+// ServiceCtrlHandler is called when the service is being controlled by Windows
+void WINAPI ServiceCtrlHandler(DWORD ctrlCode) {
+    switch (ctrlCode) {
+        case SERVICE_CONTROL_STOP:
+        case SERVICE_CONTROL_SHUTDOWN:
+            serviceStatus.dwCurrentState = SERVICE_STOP_PENDING;
+            serviceStatus.dwWin32ExitCode = 0;
+            serviceStatus.dwCheckPoint = 0;
+            serviceStatus.dwWaitHint = 0;
+            SetServiceStatus(serviceStatusHandle, &serviceStatus);
+            GoStop();
```

Comment:
my worry is how do we wait for processes to finish work, flush things

Reply:
This flow is just setting it to stop pending and calling `GoStop`. `GoStop` is non-blocking and will return instantly. Its job is to just cancel the context and return.

I have actually updated this flow in my latest commit and I am seeing integration tests success with the latest changes.